### PR TITLE
perf(@ngtools/webpack): reduce rebuild performance by typechecking more

### PR DESCRIPTION
**Pull request by hansl**
We got the TypeScript performance down a lot by not type checking every files on rebuild, but this has an issue where typings can be wrong in files that depends on us. This PR alleviate the problem by type checking all files that depends on the changed file.

Note that even if this PR reduces performance, it's marginal by less than 5% (although it depends on the depth of the change in the dependency graph).

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -55,7 +55,7 @@
     "ember-cli-string-utils": "^1.0.0",
     "enhanced-resolve": "^2.3.0",
     "exists-sync": "0.0.3",
-    "extract-text-webpack-plugin": "^2.0.0-rc.1",
+    "extract-text-webpack-plugin": "^2.0.0-beta.5",
```
Comment (filipesilva) [inline package.json]: Why does this need to be changed?

File: packages/@ngtools/webpack/src/webpack.ts [hunk 2]
```diff
@@ -23,3 +23,29 @@ export interface ResolverPlugin extends Tapable {
   join(relativePath: string, innerRequest: Request): Request;
 }
 
+export interface LoaderCallback {
```
Comment (filipesilva) [inline packages/@ngtools/webpack/src/webpack.ts]: Wouldn't it be better to use the existing Webpack typings?

File: packages/@ngtools/webpack/src/loader.ts [hunk 3]
```diff
@@ -201,6 +202,15 @@ export function ngcLoader(source: string) {
       })
       .then(() => {
         if (plugin.typeCheck) {
+          // Check all diagnostics from this and reverse dependencies also.
+          if (!plugin.firstRun) {
+            this._module.reasons
+              .filter(reason => reason.module && reason.module instanceof NormalModule)
+              .forEach(reason => plugin.diagnose(reason.module.resource));
```
Comment (filipesilva) [inline packages/@ngtools/webpack/src/loader.ts]: Will this also get reverse dependencies of reverse dependencies? e.g. `index.ts` re-exporting classes.

File: package.json [hunk 4]
```diff
@@ -130,7 +130,7 @@
     "@types/fs-extra": "^0.0.31",
     "@types/glob": "^5.0.29",
     "@types/jasmine": "^2.2.32",
-    "@types/lodash": "4.14.50",
+    "@types/lodash": "4.14.43",
```
Comment (cacogr) [inline package.json]: `4.14.50` was pinned in https://github.com/angular/angular-cli/pull/4260 recently merged

Comment (hansl) [inline package.json]: Done. that was a merge conflict.

File: packages/@ngtools/webpack/src/loader.ts (hunk 3 above)
Comment (hansl) [inline packages/@ngtools/webpack/src/loader.ts]: Yeah I was doing that, and forgot to put it back, my bad. Thanks for catching it!

File: package.json (hunk 1 above)
Comment (hansl) [inline package.json]: Merge conflict, I'm reverting this.

File: packages/@ngtools/webpack/src/webpack.ts (hunk 2 above)
Comment (hansl) [inline packages/@ngtools/webpack/src/webpack.ts]: The `@types/webpack` package? It's super incomplete and fits more for using webpack than developing a plugin. I'm still looking for a good webpack typings from the webpack team itself.

**Pull request merged**